### PR TITLE
[dtensor][1/N] add __hash__ to device_mesh and dtensor_spec

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92126
* #90735
* #90734
* #90733
* #90732
* #91785
* __->__ #90731

This PR adds __hash__ to device_mesh and dtensor_spec to allow
things like dict indexing

### CODE REVIEW

**Review by XilunWu — APPROVED** [review]
Overall LGTM. Just 1 nit and some questions. Thx Wanchao!

File: test/distributed/_tensor/test_device_mesh.py [hunk 1]
```diff
@@ -143,6 +143,17 @@ def test_device_mesh_nd(self):
                 if self.rank in ranks:
                     self.assertEqual(global_ranks, ranks.tolist())
 
+    @with_comms
+    def test_device_mesh_hash(self):
+        mesh_tensor_2d = torch.arange(8).reshape(4, 2)
+        mesh = DeviceMesh(self.device_type, mesh_tensor_2d)
+        mesh2 = DeviceMesh(self.device_type, mesh_tensor_2d)
+        self.assertEqual(hash(mesh), hash(mesh2))
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply test/distributed/_tensor/test_device_mesh.py:151]: yeah good catch! Actually hash should respect equality so in this case mesh and mesh2 should not hash to the same value, let me update

File: test/distributed/_tensor/test_common_rules.py [hunk 2]
```diff
@@ -143,9 +143,7 @@ def test_einop_linearity(self):
         )
         mesh = DeviceMesh(self.device_type, mesh_shape)
 
-        mm_func_schema = parse_schema(
-            "aten::mm(Tensor self, Tensor mat2) -> Tensor"
-        )
+        mm_func_schema = parse_schema("aten::mm(Tensor self, Tensor mat2) -> Tensor")
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply test/distributed/_tensor/test_common_rules.py:146]: Yep we should, some format discrepency from the two code base, let me see if we can make the default formatting applied

File: torch/distributed/_tensor/device_mesh.py [hunk 3]
```diff
@@ -105,7 +105,7 @@ def __init__(
     ) -> None:
         self.device_type = device_type
         self.mesh = (
-            mesh.detach()
+            mesh
```
[the remark this answers is not in the document]
Comment (XilunWu) [reply torch/distributed/_tensor/device_mesh.py:108]: then doesn't the original `mesh.detach()` remove `mesh` from the graph? Why should we change it to `mesh` instead?

[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/distributed/_tensor/device_mesh.py]: I think detach was there because I want to avoid user modifying the mesh tensor. Removed it because of only hashing the mesh tensor, but since we should force equality, mesh + id should be hashed, so I added the detach() call back :)

File: torch/distributed/_tensor/placement_types.py [hunk 4]
```diff
@@ -270,6 +293,17 @@ def __post_init__(self) -> None:
         if self.ndim == -1:
             self.ndim = len(self.shape)
 
+    def __hash__(self) -> int:
+        return hash((self.mesh, tuple(self.placements), self.shape))
```
Comment (fduwjj) [inline torch/distributed/_tensor/placement_types.py:297]: Do we also want to include "ndim"?

**Review by fduwjj — APPROVED** [review]
Thanks for pushing this to prod.

File: torch/distributed/_tensor/placement_types.py (hunk 4 above)
Comment (wanchaol) [reply torch/distributed/_tensor/placement_types.py:297]: probably no, `ndim` is inferred from shape, so it represents the same thing and should hash to the same. Also we might want to get rid of ndim and use shape uniformly.
